vcd/config_test.go

```diff
@@ -314,9 +323,155 @@ func TestMain(m *testing.M) {
+
 	// TODO: cleanup leftovers
 	os.Exit(exitCode)
 }
+
+func createSuiteCatalogAndItem(config TestConfig) {
+	fmt.Printf("Creating resources for test suite...\n")
+	vcdClient, err := getTestVCDFromJson(config)
+	if vcdClient == nil || err != nil {
+		panic(err)
+	}
+	err = vcdClient.Authenticate(config.Provider.User, config.Provider.Password, config.Provider.SysOrg)
+	if err != nil {
+		panic(err)
+	}
+
+	org, err := govcd.GetOrgByName(vcdClient, config.VCD.Org)
+	if err != nil || org == (govcd.Org{}) {
+		panic(err)
+	}
+
+	var catalog govcd.Catalog
+
+	if testConfig.VCD.Catalog.Name == "" {
+		fmt.Printf("Creating catalog for test suite...\n")
+		catalog, err = org.CreateCatalog(testSuiteCatalogName, "Test suite purpose")
+		if err != nil || catalog == (govcd.Catalog{}) {
+			panic(err)
+		}
+		fmt.Printf("Catalog created successfuly\n")
+
+	} else {
+		fmt.Printf("Skiping catalog creation - found preconfigured one: %s \n", testConfig.VCD.Catalog.Name)
+
+		catalog, err = org.FindCatalog(testConfig.VCD.Catalog.Name)
+		if err != nil || catalog == (govcd.Catalog{}) {
+			fmt.Printf("Preconfigured catalog wasn't found \n")
+			panic(err)
+		}
+
+		fmt.Printf("Catalog found successfuly\n")
+		testSuiteCatalogName = testConfig.VCD.Catalog.Name
+	}
+
+	if testConfig.VCD.Catalog.CatalogItem == "" {
+		fmt.Printf("Creating catalog item for test suite...\n")
+		task, err := catalog.UploadOvf("../test-resources/"+config.Ova.OvaTestFileName, testSuiteCatalogOVAItem, "Test suite purpose", 20*1024*1024)
+		if err != nil {
+			fmt.Errorf("error uploading new catalog item: %#v", err)
+			panic(err)
+		}
+
+		err = task.ShowUploadProgress()
+		if err != nil {
+			fmt.Errorf("error waiting from task to complete: %+v", err)
+			panic(err)
+		}
+
+		err = task.WaitTaskCompletion()
+		if err != nil {
+			fmt.Errorf("error waiting from task to complete: %+v", err)
+			panic(err)
+		}
+
+		fmt.Printf("Catalog item created successfuly\n")
+
+	} else {
+		fmt.Printf("Skiping catalog item creation - found preconfigured one: %s \n", testConfig.VCD.Catalog.CatalogItem)
+
+		item, err := catalog.FindCatalogItem(testConfig.VCD.Catalog.CatalogItem)
+		if err != nil && item != (govcd.CatalogItem{}) {
+			fmt.Printf("Preconfigured catalog item wasn't found \n")
+			panic(err)
+		}
+		fmt.Printf("Catalog item found successfuly\n")
+		testSuiteCatalogOVAItem = testConfig.VCD.Catalog.CatalogItem
+	}
+
+}
+
+// Creates a VCDClient based on the endpoint given in the TestConfig argument.
+// TestConfig struct can be obtained by calling GetConfigStruct. Throws an error
+// if endpoint given is not a valid url.
+func getTestVCDFromJson(testConfig TestConfig) (*govcd.VCDClient, error) {
+	configUrl, err := url.ParseRequestURI(testConfig.Provider.Url)
+	if err != nil {
+		return &govcd.VCDClient{}, fmt.Errorf("could not parse Url: %s", err)
+	}
+	vcdClient := govcd.NewVCDClient(*configUrl, true)
+	return vcdClient, nil
+}
+
+func destroySuiteCatalogAndItem(config TestConfig) {
```

Comment:
Should we have an option for keeping such resources for another test?

Reply:
This destroy function called after all tests are finished.

Reply:
I mean: to use it with the next test suite run.

run n. 1: create default catalog and item. If we have an option to preserve them, they are not deleted.
run n.2: if the default catalog and item already exist, we reuse them, without another download/upload setup operation

Currently, if the catalog already exists (for a failure of the previous test) we get an unrecoverable error.

Reply:
During the call we agreed on some additional behaviour. Now json configuration holds `preserve` which means not to destroy test suite created resources and on next run will be reused. TEST_SUITE_CLEANUP env variable will allow to force resource deletion before testsuite running - in case aborted or errors with previous test run.